src/Algebra/Graph/Bipartite/AdjacencyMap.hs

```diff
@@ -0,0 +1,431 @@
+-- Complexity: /O(m log n)/.
+--
+-- @
+-- 'toBipartite ('Algebra.Graph.AdjacencyMap.empty')                                                           == 'empty'
+-- 'leftAdjacencyMap' (toBipartite ('Algebra.Graph.AdjacencyMap.edge' (Left 1) (Right 2)))                       == Map.'Map.singleton' 1 (Set.'Set.singleton' 2)
+-- toBipartite ('Algebra.Graph.AdjacencyMap.edge' (Left 1) (Left 2))                                           == 'empty'
+-- 'rightAdjacencyMap' (toBipartite ('Algebra.Graph.AdjacencyMap.edges' [(Left 1, Right 1), (Left 1, Right 2)])) == Map.'Map.fromAscList' [(1, Set.'Set.singleton' 1), (2, Set.'Set.singleton' 1)]
+-- @
+toBipartite :: (Ord a, Ord b) => AM.AdjacencyMap (Either a b) -> AdjacencyMap a b
+toBipartite m = BAM (Map.fromAscList [ (u, setRights vs) | (Left  u, vs) <- Map.toAscList (AM.adjacencyMap $ addReverseEdges m)])
+                         (Map.fromAscList [ (u, setLefts  vs) | (Right u, vs) <- Map.toAscList (AM.adjacencyMap $ addReverseEdges m)])
+    where
+        setRights = Set.fromAscList . rights . Set.toAscList
+        setLefts  = Set.fromAscList . lefts  . Set.toAscList
+
+-- | Constructs an 'Algrebra.Graph.AdjacencyMap' from a bipartite
+-- 'AdjacencyMap'.
+-- Complexity: /O(m log n)/.
+--
+-- @
+-- fromBipartite 'empty'          == 'Algebra.Graph.AdjacencyMap.empty'
+-- fromBipartite ('leftVertex' 1) == 'Algebra.Graph.AdjacencyMap.vertex' (Left 1)
+-- fromBipartite ('edge' 1 2)     == 'Algebra.Graph.AdjacencyMap.edges' [(Left 1, Right 2), (Right 2, Left 1)]
+-- @
+fromBipartite :: (Ord a, Ord b) => AdjacencyMap a b -> AM.AdjacencyMap (Either a b)
+fromBipartite (BAM lr rl) = AM.overlays $
+       [ AM.edges [ (Left u, Right v) | (u, vs) <- Map.toAscList lr, v <- Set.toAscList vs ]
+       , AM.edges [ (Right u, Left v) | (u, vs) <- Map.toAscList rl, v <- Set.toAscList vs ]
+       , AM.vertices $ Left  <$> Map.keys lr
+       , AM.vertices $ Right <$> Map.keys rl ]
+
+-- | Constructs a bipartite 'AdjacencyMap' from a 'Algebra.Graph.Graph' with
+-- given part identifiers, adding all needed edges to make the graph undirected
+-- and removing all edges inside one part.
+-- Complexity: /O(m log n)/.
+--
+-- @
+-- 'leftAdjacencyMap' (fromGraph ('Algebra.Graph.empty'))                                         == Map.'Map.empty'
+-- 'leftAdjacencyMap' (fromGraph ('Algebra.Graph.edge' (Left 1) (Right 2)))                       == Map.'Map.singleton' 1 (Set.'Set.singleton' 2)
+-- 'leftAdjacencyMap' (fromGraph ('Algebra.Graph.edge' (Left 1) (Left 2)))                        == Map.'Map.empty'
+-- 'rightAdjacencyMap' (fromGraph ('Algebra.Graph.edges' [(Left 1, Right 1), (Left 1, Right 2)])) == Map.'Map.fromAscList' [(1, Set.'Set.singleton' 1), (2, Set.'Set.singleton' 1)]
+-- @
+fromGraph :: (Ord a, Ord b) => G.Graph (Either a b) -> AdjacencyMap a b
+fromGraph = toBipartite . (G.foldg AM.empty AM.vertex AM.overlay AM.connect)
+
+-- | Constructs the /empty graph/.
```

Comment:
Please reorder all function definitions in this module according to the export list. This helps navigating it.

Reply:
Also, let's use "Construct" instead of "Constructs" throughout the module.